### ISSUE TITLE
Sprout crash at high traffic load when configured with a CFS TAS.

### ISSUE THREAD

**Issue by plfrancis** [Metaswitch/sprout#981]
When attempting to run 100+ CPS on a single Sprout instance configured with a CFS TAS, Sprout crashes when trying to build an ACR. This behavior was triggered by enabling billing headers in the TAS. At first glance it would appear that  store_charging_addresses() may need reentrancy protection for modifications to the _ccfs and _ecfs lists. The following is the backtrace for the offending thread (a diagnostic bundle was also collected).

```
(gdb) bt
#0  0x00007fdd0b480425 in __GI_raise (sig=<optimized out>) at ../nptl/sysdeps/unix/sysv/linux/raise.c:64
#1  0x00007fdd0b483b8b in __GI_abort () at abort.c:91
#2  0x00000000005a8e7c in exception_handler (sig=6) at main.cpp:841
#3  <signal handler called>
#4  0x00007fdd0b480425 in __GI_raise (sig=<optimized out>) at ../nptl/sysdeps/unix/sysv/linux/raise.c:64
#5  0x00007fdd0b483b8b in __GI_abort () at abort.c:91
#6  0x00007fdd0b4be39e in __libc_message (do_abort=2, fmt=0x7fdd0b5c8748 "*** glibc detected *** %s: %s: 0x%s ***\n")
    at ../sysdeps/unix/sysv/linux/libc_fatal.c:201
#7  0x00007fdd0b4c8b96 in malloc_printerr (action=3, str=0x7fdd0b5c8938 "double free or corruption (fasttop)", ptr=<optimized out>) at malloc.c:5039
#8  0x00000000004e6177 in _M_dispose (__a=..., this=<optimized out>) at /usr/include/c++/4.6/bits/basic_string.h:244
#9  ~basic_string (this=0x7fdc9467bf00, __in_chrg=<optimized out>) at /usr/include/c++/4.6/bits/basic_string.h:534
#10 ~_List_node (this=0x7fdc9467bef0, __in_chrg=<optimized out>) at /usr/include/c++/4.6/bits/stl_list.h:105
#11 destroy (__p=0x7fdc9467bef0, this=<optimized out>) at /usr/include/c++/4.6/ext/new_allocator.h:118
#12 std::_List_base<std::string, std::allocator<std::string> >::_M_clear (this=0x7fdc946797f0) at /usr/include/c++/4.6/bits/list.tcc:76
#13 0x000000000057e9b1 in clear (this=0x7fdc946797f0) at /usr/include/c++/4.6/bits/stl_list.h:1197
#14 RalfACR::store_charging_addresses (this=0x7fdc946797d0, msg=<optimized out>) at acr.cpp:1400
#15 0x000000000057ed3d in RalfACR::rx_response (this=0x7fdc946797d0, rsp=0x7fdc846c2490, timestamp=...) at acr.cpp:525
#16 0x00007fdcf83d811b in SCSCFSproutletTsx::on_rx_response (this=0x7fdc943089b0, rsp=0x7fdc846c2490, fork_id=0) at scscfsproutlet.cpp:480
#17 0x000000000059fc10 in SproutletWrapper::rx_response (this=0x7fdc9475c790, rsp=0x7fdc846c1f28, fork_id=0) at sproutletproxy.cpp:1417
#18 0x00000000005a0d56 in SproutletProxy::UASTsx::on_new_client_response (this=0x7fdc946241c0, uac_tsx=<optimized out>, rsp=0x7fdc846c1f28)
    at sproutletproxy.cpp:587
#19 0x000000000057b05f in BasicProxy::UACTsx::on_tsx_state (this=0x7fdc94517f60, event=0x7fdcbfda1bf0) at basicproxy.cpp:1871
#20 0x00000000005ec5dc in tsx_set_state (tsx=0x7fdc94516558, state=PJSIP_TSX_STATE_PROCEEDING, event_src_type=PJSIP_EVENT_RX_MSG, event_src=0x7fdca49cb6d8)
    at ../src/pjsip/sip_transaction.c:1199
#21 0x00000000005efd4f in tsx_on_state_proceeding_uac (tsx=0x7fdc94516558, event=0x7fdcbfda1ce0) at ../src/pjsip/sip_transaction.c:2867
#22 0x00000000005ef1aa in tsx_on_state_calling (tsx=0x7fdc94516558, event=0x7fdcbfda1ce0) at ../src/pjsip/sip_transaction.c:2491
#23 0x00000000005ed806 in pjsip_tsx_recv_msg (tsx=0x7fdc94516558, rdata=0x7fdca49cb6d8) at ../src/pjsip/sip_transaction.c:1754
#24 0x00000000005ebb26 in mod_tsx_layer_on_rx_response (rdata=0x7fdca49cb6d8) at ../src/pjsip/sip_transaction.c:863
#25 0x00000000005d5ca4 in pjsip_endpt_process_rx_data (endpt=0x7fdd036fa0b8, rdata=0x7fdca49cb6d8, p=0x7fdcbfda1e00, p_handled=0x0)
    at ../src/pjsip/sip_endpoint.c:893
#26 0x00000000004a19f1 in worker_thread (p=<optimized out>) at stack.cpp:207
#27 0x000000000060acd0 in thread_main (param=0x283ed78) at ../src/pj/os_core_unix.c:523
#28 0x00007fdd0c023e9a in start_thread (arg=0x7fdcbfda2700) at pthread_create.c:308
#29 0x00007fdd0b53e3fd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:112
#30 0x0000000000000000 in ?? ()
(gdb)
```

**Comment by rkd-msw** [comment]
Is it possible to reproduce this with debug logging turned on? That would give a better idea of what has previously happened to that ACR (in particular, if it's previously been destroyed).

Does this only reproduce at 100+ CPS load, or is it possible to see this on individual calls where the AS adds billing headers?

**Comment by plfrancis** [comment]
This was only reproducible at high call rates, which seems to imply that the window of vulnerability is pretty small. At those rates it would not be possible to turn on debug level logging.

**Comment by mirw** [comment]
The stack is similar to #1095.  It's worth looking out for this again after that's fixed, as it could be a subtly-different instance of that.

**Comment by mirw** [comment]
I think next action is with @plfrancis to get diagnostics (e.g. SAS trace) from this failure.

**Comment by plfrancis** [comment]
This issue has not been observed in recent traffic runs/soaks. Closing.

**Comment by plfrancis** [comment]
This issue has been seen when using a standalone Sproutlet based TAS "stub" in lieu of a VoLTE TAS. As with the original write-up, the trigger seems to be related to billing headers in INVITEs from the TAS. In this case Sprout was running duplex, while a single Ralf was being used for performance measurements. The following is a backtrace from CC v7 (a diagnostic bundle was also collected).

```
(gdb) bt
#0  0x00007ffad84f4cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffad84f80d8 in __GI_abort () at abort.c:89
#2  0x00000000005bd7ca in ExceptionHandler::handle_exception (this=0x241c8f0)
    at /var/build/CC_7.0.0-WIP/cc/sprout/modules/cpp-common/src/exception_handler.cpp:73
#3  0x00000000005bdbcd in signal_handler (sig=6) at main.cpp:1015
#4  <signal handler called>
#5  0x00007ffad84f4cc9 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#6  0x00007ffad84f80d8 in __GI_abort () at abort.c:89
#7  0x00007ffad8531394 in __libc_message (do_abort=do_abort@entry=1,
    fmt=fmt@entry=0x7ffad863fb28 "*** Error in `%s': %s: 0x%s ***\n")
    at ../sysdeps/posix/libc_fatal.c:175
#8  0x00007ffad853d66e in malloc_printerr (ptr=<optimized out>,
    str=0x7ffad863fcf0 "double free or corruption (fasttop)", action=1)
    at malloc.c:4996
#9  _int_free (av=<optimized out>, p=<optimized out>, have_lock=0)
    at malloc.c:3840
#10 0x00000000004c25da in _M_dispose (__a=..., this=<optimized out>)
    at /usr/include/c++/4.8/bits/basic_string.h:249
#11 ~basic_string (this=0x7ffa581e99e0, __in_chrg=<optimized out>)
    at /usr/include/c++/4.8/bits/basic_string.h:539
---Type <return> to continue, or q <return> to quit---
#12 ~_List_node (this=0x7ffa581e99d0, __in_chrg=<optimized out>)
    at /usr/include/c++/4.8/bits/stl_list.h:106
#13 destroy<std::_List_node<std::basic_string<char> > > (this=<optimized out>,
    __p=0x7ffa581e99d0) at /usr/include/c++/4.8/ext/new_allocator.h:124
#14 std::_List_base<std::string, std::allocator<std::string> >::_M_clear (
    this=this@entry=0x7ffa58179510) at /usr/include/c++/4.8/bits/list.tcc:75
#15 0x00000000005906c6 in clear (this=0x7ffa58179510)
    at /usr/include/c++/4.8/bits/stl_list.h:1230
#16 RalfACR::store_charging_addresses (this=this@entry=0x7ffa581794f0,
    msg=msg@entry=0x7ffa48747ca0) at acr.cpp:1437
#17 0x0000000000591455 in RalfACR::rx_response (this=0x7ffa581794f0,
    rsp=0x7ffa48747ca0, timestamp=...) at acr.cpp:530
#18 0x00007ffac03d85a2 in SCSCFSproutletTsx::on_rx_response (
    this=0x7ffa585504a0, rsp=0x7ffa48747ca0, fork_id=0)
    at scscfsproutlet.cpp:551
#19 0x00000000005b3865 in SproutletWrapper::rx_response (this=0x7ffa58259a00,
    rsp=0x7ffa48747738, fork_id=0) at sproutletproxy.cpp:1425
#20 0x00000000005b5a4c in SproutletProxy::UASTsx::on_new_client_response (
    this=0x7ffa5878fdd0, uac_tsx=<optimized out>, rsp=0x7ffa48747738)
    at sproutletproxy.cpp:549
#21 0x000000000058cdae in BasicProxy::UACTsx::on_tsx_state (
    this=0x7ffa58757dd0, event=0x7ffa9f7ddad0) at basicproxy.cpp:1904
#22 0x000000000063f489 in tsx_set_state (tsx=0x7ffa5856edd8,
---Type <return> to continue, or q <return> to quit---
    state=PJSIP_TSX_STATE_PROCEEDING, event_src_type=PJSIP_EVENT_RX_MSG,
    event_src=0x7ffa78637f88) at ../src/pjsip/sip_transaction.c:1199
#23 0x0000000000642b72 in tsx_on_state_proceeding_uac (tsx=0x7ffa5856edd8,
    event=0x7ffa9f7ddbb0) at ../src/pjsip/sip_transaction.c:2867
#24 0x000000000064207f in tsx_on_state_calling (tsx=0x7ffa5856edd8,
    event=0x7ffa9f7ddbb0) at ../src/pjsip/sip_transaction.c:2491
#25 0x000000000064069b in pjsip_tsx_recv_msg (tsx=0x7ffa5856edd8,
    rdata=0x7ffa78637f88) at ../src/pjsip/sip_transaction.c:1754
#26 0x000000000063e9b5 in mod_tsx_layer_on_rx_response (rdata=0x7ffa78637f88)
    at ../src/pjsip/sip_transaction.c:863
#27 0x000000000062822f in pjsip_endpt_process_rx_data (endpt=0x7ffacaced0b8,
    rdata=0x7ffa78637f88, p=0x7ffa9f7ddd20, p_handled=0x0)
    at ../src/pjsip/sip_endpoint.c:893
#28 0x00000000005bae5d in worker_thread (p=<optimized out>)
    at thread_dispatcher.cpp:153
#29 0x00000000006a4fae in thread_main (param=0x2b55848)
    at ../src/pj/os_core_unix.c:523
#30 0x00007ffad90ab182 in start_thread (arg=0x7ffa9f7de700)
    at pthread_create.c:312
#31 0x00007ffad85b847d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```.

**Comment by JimBunch** [comment]
Marked as critical because Sprout crashes which is call processing service affecting.

**Comment by mirw** [comment]
We see from the stack that we're crashing because we're clearing the `_ccfs` `std::vector` in a `RalfACR`.  There's no good reason for this to crash... unless the `RalfACR` has already been destroyed.

I wondered if this might be related to #1111, which changed the lifespan of ACRs.  However, it looks like that fix might have been in this release.  (It's quite hard to tell - it looks as though the Sprout build that was running predates the merge by a couple of hours, but the fact that it was built on the same day the fix was merged suggests that maybe this was intended to be the same release.)

How easy is it to repro this?  We can do more digging, but a SAS flow (which should now be identified in the diags bundle) would make life much easier!  Please let me know if that's difficult.

**Comment by sathiyan-sivathas** [comment]
Next action is to collect a SAS trace of the sprout crash at 100+ CPS.  We're blocked on that.

**Comment by chris-elford-metaswitch** [comment]
@rkd-msw - It looks like the next step on this one is to gather diags. Is that something that you can run with?

**Comment by rkd-msw** [comment]
As discussed, next action is with you to decide on how urgent this is to investigate and fix.

**Comment by chris-elford-metaswitch** [comment]
Given that this we've seen this with a couple of TAS's, I'd like to understand whether this is something that we are likely to hit whenever Ralf is under stress. 

Rob - you offered to run some stress tests. If we can't reproduce this, let's close it down again.

**Comment by rkd-msw** [comment]
Yep, it reproduces (by adding a P-Charging-Function-Addresses header and running stress through a Mangelwurzel AS):

```
Thread 198 (Thread 0x7fd128ff9700 (LWP 20510)):
#0  0x00007fd162544b99 in waitpid () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007fd1624ca2e2 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x0000000000515063 in Logger::backtrace (this=0x24c9d80, data=<optimized out>) at ../modules/cpp-common/src/logger.cpp:318
#3  0x00000000005d53cd in Log::backtrace (fmt=fmt@entry=0x828018 "Signal %d caught") at ../modules/cpp-common/src/log.cpp:185
#4  0x000000000063e4ec in signal_handler (sig=6) at main.cpp:1268
#5  <signal handler called>
#6  0x00007fd1624bacc9 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#7  0x00007fd1624be0d8 in abort () from /lib/x86_64-linux-gnu/libc.so.6
#8  0x00007fd1624f7394 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#9  0x00007fd16250366e in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#10 0x000000000051da00 in ~basic_string (this=0x7fd0d432d130, __in_chrg=<optimized out>) at /usr/include/c++/4.8/bits/basic_string.h:539
#11 ~_List_node (this=0x7fd0d432d120, __in_chrg=<optimized out>) at /usr/include/c++/4.8/bits/stl_list.h:106
#12 destroy<std::_List_node<std::basic_string<char> > > (this=<optimized out>, __p=0x7fd0d432d120) at /usr/include/c++/4.8/ext/new_allocator.h:124
#13 std::_List_base<std::string, std::allocator<std::string> >::_M_clear (this=this@entry=0x7fd0d43f02a8) at /usr/include/c++/4.8/bits/list.tcc:75
#14 0x00000000005f2486 in clear (this=0x7fd0d43f02a8) at /usr/include/c++/4.8/bits/stl_list.h:1230
#15 RalfACR::store_charging_addresses (this=this@entry=0x7fd0d43f0280, msg=msg@entry=0x7fd14c3c34b0) at acr.cpp:1483
#16 0x00000000005f2be7 in RalfACR::rx_request (this=0x7fd0d43f0280, req=0x7fd14c3c34b0, timestamp=...) at acr.cpp:446
#17 0x00007fd1582efe6f in SCSCFSproutletTsx::on_rx_initial_request (this=0x7fd14c2c56f0, req=0x7fd14c3c34b0) at scscfsproutlet.cpp:486
#18 0x0000000000611633 in SproutletWrapper::rx_request (this=0x7fd14c1bd1d0, req=0x7fd14c1bd468) at sproutletproxy.cpp:1583
#19 0x00000000006121fd in SproutletProxy::UASTsx::process_tsx_request (this=0x7fd14c17c890, rdata=<optimized out>) at sproutletproxy.cpp:592
#20 0x00000000005eeedd in BasicProxy::on_tsx_request (this=0x3332750, rdata=0x7fd1544181a8) at basicproxy.cpp:283
#21 0x00000000005ed2ce in BasicProxy::on_rx_request (this=0x3332750, rdata=0x7fd1544181a8) at basicproxy.cpp:93
#22 0x00000000007509df in pjsip_endpt_process_rx_data (endpt=0x7fd151ceb0b8, rdata=0x7fd1544181a8, p=0x7fd128ff8d10, p_handled=0x0) at ../src/pjsip/sip_endpoint.c:885
#23 0x000000000061896d in worker_thread (p=<optimized out>) at thread_dispatcher.cpp:154
#24 0x00000000007cd8cd in thread_main (param=0x3418370) at ../src/pj/os_core_unix.c:523
#25 0x00007fd1639c2182 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#26 0x00007fd16257e47d in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

I've attached a SAS trace and the diags package.

**Comment by rkd-msw** [comment]
@chris-elford-metaswitch - should this be major?

**Comment by chris-elford-metaswitch** [comment]
Yes

**Comment by rkd-msw** [comment]
I can't reproduce this under memcheck or helgrind (and helgrind doesn't warn about anything related to RalfACR) - I think the delay introduced by valgrind throws off the timing windows.

Worth temporarily upping the severity of the log at https://github.com/Metaswitch/sprout/blob/dev/src/acr.cpp#L54 to see if this ACR has already been destroyed?

**Comment by sebrexmetaswitch** [comment]
I added a bunch of extra logging in this area and repro'd the problem.

I think it's being caused by more than one thread accessing the RalfACR at once. The RalfACR is owned by the AsChain object, which is accessed by potentially several SCSCFSproutletTsx objects. (If there is an AS involved in the call which acts as a back to back user agent, then we'll have 2 transactions that access the same AsChain, and these can be handled by different threads simulataneously as we lock at the Tsx level.) Each Tsx object informs the acr of every (non-100) SIP message that it sends/receives, and most of these involve changing state on the acr.

In the case of this crash, I think 2 threads are clearing the _ccfs std::list at the same time.

**Comment by rkd-msw** [comment]
Suspect it doesn't need to act as a B2BUA - any AS will create a new transaction. (B2BUAs create separate dialogs.)

**Comment by sebrexmetaswitch** [comment]
This is now merged, so should be in the next nightly build. Back to you to retest.

**Comment by sebrexmetaswitch** [comment]
(I think - I've just noticed that this wasn't raised by you, so pass it back to me if you think the next step's not with you)

**Comment by rkd-msw** [comment]
Yes, I'll retest.

**Comment by rkd-msw** [comment]
Confirmed as fixed (on 10.0.0-170223-~255.00.0-170223.020355).
